src/oid4vp/permission_request.rs

```diff
@@ -0,0 +1,204 @@
+    /// public, however, it is likely that the `from_definition` method will be used to construct
+    /// requested fields from a presentation definition.
+    ///
+    /// See [RequestedField::from_definition] to return a vector of requested fields
+    /// according to a presentation definition.
+    pub fn new(
+        name: String,
+        required: bool,
+        retained: bool,
+        purpose: Option<String>,
+        input_descriptor_id: String,
+    ) -> Arc<Self> {
+        Arc::new(Self {
+            id: Uuid::new_v4(),
+            name,
+            required,
+            retained,
+            purpose,
+            input_descriptor_id,
+        })
+    }
+
+    /// Return the unique ID for the request field.
+    pub fn id(&self) -> Uuid {
+        self.id
+    }
+
+    /// Return the input descriptor id.
+    pub fn input_descriptor_id(&self) -> String {
+        self.input_descriptor_id.clone()
+    }
+}
+
+/// Public methods for the RequestedField struct.
+#[uniffi::export]
+impl RequestedField {
+    /// Return the field name
+    pub fn name(&self) -> String {
+        self.name.clone()
+    }
+
+    /// Return the field required status
+    pub fn required(&self) -> bool {
+        self.required
+    }
+
+    /// Return the field retained status
+    pub fn retained(&self) -> bool {
+        self.retained
+    }
+
+    /// Return the purpose of the requested field.
+    pub fn purpose(&self) -> Option<String> {
+        self.purpose.clone()
+    }
+}
+
+#[derive(Debug, Clone, uniffi::Object)]
+pub struct PermissionRequest {
+    definition: PresentationDefinition,
+    credentials: Vec<Arc<ParsedCredential>>,
+}
+
+impl PermissionRequest {
+    pub fn new(
+        definition: PresentationDefinition,
+        credentials: Vec<Arc<ParsedCredential>>,
+    ) -> Arc<Self> {
+        Arc::new(Self {
+            definition,
+            credentials,
+        })
+    }
+}
+
+#[uniffi::export]
+impl PermissionRequest {
+    /// Return the filtered list of credentials that matched
+    /// the presentation definition.
+    pub fn credentials(&self) -> Vec<Arc<ParsedCredential>> {
+        self.credentials.clone()
+    }
+
+    /// Return the requested fields for a given credential id.
+    pub fn requested_fields(&self) -> Vec<Arc<RequestedField>> {
+        self.definition
+            .input_descriptors()
+            .iter()
+            .flat_map(|descriptor| {
+                descriptor.constraints().fields().iter().map(|field| {
+                    let purpose = field.purpose().map(ToOwned::to_owned);
+                    let name = field
+                        .name()
+                        .map(ToOwned::to_owned)
+                        // TODO: Add an "unknown field" if the name is not provided.
+                        // Consider skipping or erroring on unknown fields.
+                        .unwrap_or_default();
```

Comment:
I think it would be better to use the selected credentials, so you can apply the JSONPath to the selected credentials and get the actual field name from that credential. We have no guarantee that the field name that is being requested actually has any bearing on what data is actually being requested, so this could end up being an attack vector.